# Customer Retention & Churn Analysis

## Project Objective

The objective of this project is to analyze customer churn behavior, identify factors contributing to customer attrition, measure retention performance, and provide actionable recommendations to improve customer retention and reduce revenue loss.

## Dataset

Source: IBM Telco Customer Churn Dataset

This dataset contains customer demographics, subscription details, revenue information, churn status, and customer lifetime value metrics.

## Step 1: Import Required Libraries

### Purpose

Import the libraries required for data manipulation and analysis.

### Libraries Used

- Pandas: Data loading and analysis
- NumPy: Numerical operations

### Business Relevance

These libraries help us explore customer data and calculate business KPIs.

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

## Step 2: Load Dataset

### Purpose

Load the customer churn dataset into a Pandas DataFrame.

### Business Relevance

This dataset represents the company's customer base and will be used for retention and churn analysis.

In [11]:
df = pd.read_excel("../data/raw/Telco_customer_churn.xlsx")


## Step 3: Preview Dataset

### Business Question

What does the customer data look like?

### Purpose

Display the first few rows to verify that the dataset has been loaded correctly.

In [12]:
df.head()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,Senior Citizen,Partner,Dependents,Tenure Months,Phone Service,Multiple Lines,Internet Service,Online Security,Online Backup,Device Protection,Tech Support,Streaming TV,Streaming Movies,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,No,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,No,No,Yes,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,No,No,Yes,8,Yes,Yes,Fiber optic,No,No,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,No,Yes,Yes,28,Yes,Yes,Fiber optic,No,No,Yes,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,No,No,Yes,49,Yes,Yes,Fiber optic,No,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices


## Step 4: Dataset Size

### Business Question

How many customers and attributes are available for analysis?

### Purpose

Identify the number of rows and columns present in the dataset.

In [13]:
df.shape

(7043, 33)

## Step 5: Review Column Names

### Business Question

What customer information is available in the dataset?

### Purpose

Review all available columns before analysis.

In [14]:
df.columns

Index(['CustomerID', 'Count', 'Country', 'State', 'City', 'Zip Code',
       'Lat Long', 'Latitude', 'Longitude', 'Gender', 'Senior Citizen',
       'Partner', 'Dependents', 'Tenure Months', 'Phone Service',
       'Multiple Lines', 'Internet Service', 'Online Security',
       'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV',
       'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method',
       'Monthly Charges', 'Total Charges', 'Churn Label', 'Churn Value',
       'Churn Score', 'CLTV', 'Churn Reason'],
      dtype='str')

## Step 6: Dataset Information

### Business Question

Are the columns stored using appropriate data types?

### Purpose

Review data types, non-null counts, and overall dataset structure.

In [15]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   str    
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   str    
 3   State              7043 non-null   str    
 4   City               7043 non-null   str    
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   str    
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   str    
 10  Senior Citizen     7043 non-null   str    
 11  Partner            7043 non-null   str    
 12  Dependents         7043 non-null   str    
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   str    
 15  Multiple Lines     7043 non-null   str    
 16  Internet Service   7043 non-null   

## Step 7: Statistical Summary

### Business Question

What are the key characteristics of the numerical variables?

### Purpose

Understand the distribution of revenue, tenure, CLTV, and other numeric fields.

In [16]:
df.describe()

,Count,Zip Code,Latitude,Longitude,Tenure Months,Monthly Charges,Churn Value,Churn Score,CLTV
count,7043.0,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000
mean,1.0,93521.964646,36.282441,-119.798880,32.371149,64.761692,0.265370,58.699418,4400.295755
std,0.0,1865.794555,2.455723,2.157889,24.559481,30.090047,0.441561,21.525131,1183.057152
min,1.0,90001.000000,32.555828,-124.301372,0.000000,18.250000,0.000000,5.000000,2003.000000
25%,1.0,92102.000000,34.030915,-121.815412,9.000000,35.500000,0.000000,40.000000,3469.000000
50%,1.0,93552.000000,36.391777,-119.730885,29.000000,70.350000,0.000000,61.000000,4527.000000
75%,1.0,95351.000000,38.224869,-118.043237,55.000000,89.850000,1.000000,75.000000,5380.500000
max,1.0,96161.000000,41.962127,-114.192901,72.000000,118.750000,1.000000,100.000000,6500.000000


## Step 8: Missing Value Assessment

### Business Question

Does the dataset contain incomplete customer information?

### Purpose

Identify missing values across all columns.

In [41]:
df.isnull().sum()

CustomerID              0
Count                   0
Country                 0
State                   0
City                    0
Zip Code                0
Lat Long                0
Latitude                0
Longitude               0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Internet Service        0
Online Security         0
Online Backup           0
Device Protection       0
Tech Support            0
Streaming TV            0
Streaming Movies        0
Contract                0
Paperless Billing       0
Payment Method          0
Monthly Charges         0
Total Charges           0
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
Churn Reason         5174
dtype: int64

## Step 9: Columns with Missing Values

### Business Question

Which specific columns require data quality review?

### Purpose

Display only the columns containing missing values.

In [18]:
missing = df.isnull().sum()

missing[missing > 0]

Churn Reason    5174
dtype: int64

## Step 10: Duplicate Record Check

### Business Question

Are any customer records duplicated?

### Purpose

Identify duplicate rows that could distort KPI calculations.

In [19]:
df.duplicated().sum()

np.int64(0)

## Step 11: Unique Customer Validation

### Business Question

How many unique customers exist in the dataset?

### Purpose

Verify customer uniqueness using CustomerID.

In [20]:
df['CustomerID'].nunique()

7043

## Step 12: Total Records Validation

### Business Question

Do the total records match the total unique customers?

### Purpose

Confirm that each row represents a unique customer.

In [21]:
len(df)

7043

## Step 13: Data Type Review

### Business Question

Which columns are categorical and which are numerical?

### Purpose

Review data types for future analysis.

In [22]:
df.dtypes

CustomerID               str
Count                  int64
Country                  str
State                    str
City                     str
Zip Code               int64
Lat Long                 str
Latitude             float64
Longitude            float64
Gender                   str
Senior Citizen           str
Partner                  str
Dependents               str
Tenure Months          int64
Phone Service            str
Multiple Lines           str
Internet Service         str
Online Security          str
Online Backup            str
Device Protection        str
Tech Support             str
Streaming TV             str
Streaming Movies         str
Contract                 str
Paperless Billing        str
Payment Method           str
Monthly Charges      float64
Total Charges         object
Churn Label              str
Churn Value            int64
Churn Score            int64
CLTV                   int64
Churn Reason             str
dtype: object

## Step 14: Numerical Features

### Purpose

Identify columns that can be aggregated and used for KPI calculations.

### Examples

- Monthly Charges
- Total Charges
- CLTV
- Tenure Months

In [23]:
df.select_dtypes(include=['int64', 'float64']).columns

Index(['Count', 'Zip Code', 'Latitude', 'Longitude', 'Tenure Months',
       'Monthly Charges', 'Churn Value', 'Churn Score', 'CLTV'],
      dtype='str')

## Step 15: Categorical Features

### Purpose

Identify columns that can be used for segmentation and grouping analysis.

### Examples

- Gender
- Contract
- State
- Payment Method

In [24]:
df.select_dtypes(include='object').columns

C:\Users\jashwanth.s\AppData\Local\Temp\ipykernel_6408\3732952691.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.select_dtypes(include='object').columns


Index(['CustomerID', 'Country', 'State', 'City', 'Lat Long', 'Gender',
       'Senior Citizen', 'Partner', 'Dependents', 'Phone Service',
       'Multiple Lines', 'Internet Service', 'Online Security',
       'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV',
       'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method',
       'Total Charges', 'Churn Label', 'Churn Reason'],
      dtype='str')

## Step 16: Category Validation

### Business Question

Are category values consistent and free from data entry issues?

### Purpose

Review all unique values within categorical columns.

In [25]:
for col in df.select_dtypes(include='object').columns:
    print(f"\n----- {col} -----")
    print(df[col].unique())


----- CustomerID -----
<StringArray>
['3668-QPYBK', '9237-HQITU', '9305-CDSKC', '7892-POOKP', '0280-XJGEX',
 '4190-MFLUW', '8779-QRDMV', '1066-JKSGK', '6467-CHFZW', '8665-UTDHZ',
 ...
 '0871-OPBXW', '3605-JISKB', '9767-FFLEM', '8456-QDAVC', '7750-EYXWZ',
 '2569-WGERO', '6840-RESVB', '2234-XADUH', '4801-JZAZL', '3186-AJIEK']
Length: 7043, dtype: str

----- Country -----
<StringArray>
['United States']
Length: 1, dtype: str

----- State -----
<StringArray>
['California']
Length: 1, dtype: str

----- City -----
<StringArray>
[    'Los Angeles',   'Beverly Hills', 'Huntington Park',         'Lynwood',
  'Marina Del Rey',       'Inglewood',    'Santa Monica',        'Torrance',
        'Whittier',        'La Habra',
 ...
      'Janesville',      'Litchfield',        'Loyalton',        'Madeline',
    'Markleeville',         'Milford',         'Calpine',        'Standish',
        'Tulelake',  'Olympic Valley']
Length: 1129, dtype: str

----- Lat Long -----
<StringArray>
['33.964131, -118.2

C:\Users\jashwanth.s\AppData\Local\Temp\ipykernel_6408\1171945375.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for col in df.select_dtypes(include='object').columns:


## Step 17: Review Key Business Columns

### Business Question

What do the most important churn-related variables look like?

### Purpose

Preview customer demographics, contract details, revenue, and churn information.

In [28]:
df[['CustomerID',
    'Gender',
    'Tenure Months',
    'Contract',
    'Monthly Charges',
    'Total Charges',
    'Churn Label']].head()

,CustomerID,Gender,Tenure Months,Contract,Monthly Charges,Total Charges,Churn Label
0,3668-QPYBK,Male,2,Month-to-month,53.85,108.15,Yes
1,9237-HQITU,Female,2,Month-to-month,70.70,151.65,Yes
2,9305-CDSKC,Female,8,Month-to-month,99.65,820.5,Yes
3,7892-POOKP,Female,28,Month-to-month,104.80,3046.05,Yes
4,0280-XJGEX,Male,49,Month-to-month,103.70,5036.3,Yes


### Observation

The selected columns provide a snapshot of customer demographics, subscription details, revenue metrics, and churn status.

### Insight

These fields represent the core variables that will be used throughout KPI analysis, customer segmentation, churn analysis, and dashboard development.

## Step 18: Data Quality Summary

### Purpose

Create a consolidated data quality report.

### Metrics Reviewed

- Total Rows
- Total Columns
- Missing Values
- Duplicate Records
- Unique Customers

In [27]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

print("\nMissing Values:")
print(df.isnull().sum().sum())

print("\nDuplicate Rows:")
print(df.duplicated().sum())

print("\nUnique Customers:")
print(df['CustomerID'].nunique())

Rows: 7043
Columns: 33

Missing Values:
5174

Duplicate Rows:
0

Unique Customers:
7043


# Feature Classification

Before starting KPI calculations and exploratory analysis, we classify the dataset into:

1. Dimensions
2. Measures

This helps in:

- SQL Analysis
- Power BI Dashboard Design
- KPI Calculations
- Customer Segmentation

In [ ]:
dimensions = [
    'Gender',
    'Senior Citizen',
    'Partner',
    'Dependents',
    'Phone Service',
    'Multiple Lines',
    'Internet Service',
    'Online Security',
    'Online Backup',
    'Device Protection',
    'Tech Support',
    'Streaming TV',
    'Streaming Movies',
    'Contract',
    'Paperless Billing',
    'Payment Method',
    'State',
    'City',
    'Churn Label',
    'Churn Category',
    'Churn Reason'
]

print("Total Dimensions:", len(dimensions))

21

## Measures Identification

### Purpose

Measures are numerical variables that can be aggregated and used in KPI calculations.

Examples:

- Revenue
- CLTV
- Churn Score
- Tenure

In [ ]:
measures = [
    'Tenure Months',
    'Monthly Charges',
    'Total Charges',
    'Churn Score',
    'CLTV'
]

print("Total Measures:", len(measures))

5

# Data Understanding Summary

## Dataset Overview

The dataset contains 7,043 unique customers and 33 attributes related to demographics, geography, subscription services, revenue, customer value, and churn behavior.

## Data Quality Findings

- No duplicate records found.
- CustomerID is unique for all customers.
- Most columns contain complete information.
- Churn Reason contains missing values for active customers, which is expected and not considered a data quality issue.
- Total Charges is currently stored as a text/object field and will require conversion to numeric format before revenue analysis.

## Feature Classification

Dimensions: 21

Measures: 5

## Readiness for Analysis

The dataset is clean and suitable for customer retention, churn analysis, customer segmentation, revenue analysis, and dashboard development.